File: 2_Creational_Design_Patterns/2.04_Singleton/Java_Snippet/Singleton.java
class Printer {
    private static Printer instance;

    // Private constructor to prevent instantiation from outside
    private Printer() {
        // Initialize printer settings here
    }

    public static Printer getInstance() {
        if (instance == null) {
            instance = new Printer();
        }
        return instance;
    }

    public void printDocument(String document) {
        // Simulate printing
        System.out.println("Printing: " + document);
    }

    // Optional: to prevent cloning
    @Override
    protected Object clone() throws CloneNotSupportedException {
        throw new CloneNotSupportedException();
    }

    // Optional: to prevent deserialization
    protected Object readResolve() {
        return getInstance();
    }
}

// Usage
public class Singleton {
    public static void main(String[] args) {
        Printer printer1 = Printer.getInstance();
        Printer printer2 = Printer.getInstance();

        // Both printer1 and printer2 are references to the same object
        System.out.println(printer1 == printer2); // Output: true

        printer1.printDocument("Document 1");
        printer2.printDocument("Document 2");
    }
}

File: 2_Creational_Design_Patterns/2.04_Singleton/Python_Snippet/Singleton.py
class Printer():
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(Printer,cls).__new__(cls)
        return cls._instance
    
    def print_document(self,document):
        print("Printing ",document)
    
printer1 = Printer()
printer2 = Printer()

print(printer1 is printer2)

printer1.print_document('Document 1')
printer2.print_document('Document 2')